all.sh --keep-going

gilles-peskine-arm opened the pull request:
Keep-going mode for `tests/scripts/all.sh` so that even if one of the tests fails, the other tests run. Print a report at the end and return a nonzero status code if anything failed.

The reason to have this mode is to get complete feedback so that this script can be used for regular testing and not as a last-minute check that we don't expect to fail.

Also added options to skip yotta and armcc usage.

Backports: #1221 #1222

mazimkhan commented on tests/scripts/all.sh:
```diff
@@ -203,14 +243,78 @@ else
+        case "$*" in
+            *test|*check)
+                if [ $build_status -eq 0 ]; then
+                    record_status command make "$@"
```
```command``` is a ```bash``` built-in. Shouldn't line https://github.com/ARMmbed/mbedtls/pull/1207/files#diff-d36612fab4653d3f01f967a3c1bfc170L1 be changed to specify interpreter as ```bash```?

gilles-peskine-arm replied:
`command` is an sh builtin. It's present in all POSIX-compliant implementations of sh (including dash) and in BusyBox.

mazimkhan commented on tests/scripts/all.sh:
```diff
@@ -203,14 +243,78 @@ else
+        case "$*" in
+            *test|*check)
+                if [ $build_status -eq 0 ]; then
+                    record_status command make "$@"
```
Please confirm command ```command``` works on all supported platforms.

gilles-peskine-arm commented:
Test run on CI: https://jenkins-internal.mbed.com/job/mbedtls-release/46/

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -65,6 +66,8 @@ General options:
   -f|--force            Force the tests to overwrite any modified files.
   -k|--keep-going       Run all tests and report errors at the end.
   -m|--memory           Additional optional memory tests.
+     --armcc            Run ARM Compiler builds (on by default).
```
@mpg I'd added a positive option for `--yotta` as well. I prefer to have + and - options when practical, so you can run `all.sh --yotta` and be sure to get the yotta build even if we decide to remove it in the future.

It is inconsistent in that the older options have no inverse. Should I add them? I don't want to put too much effort into this script, I'd like to replace it by something more robust (in Python, I guess) in the medium term.

mpg replied:
I think it's OK to have positive and negative options only for things that enable/disable sets of tests, ie currently yotta and armcc, and only have one-way options for the rest.

Also, I'm a bit curious as to what problem(s) you see with this script that you think could be improved by a python rewrite?

gilles-peskine-arm replied:
What I'd really like is to have the option of running any subset of the tests. CI would run all tests (except perhaps ones that we'd declare as experimental) but it's useful to run a subset locally. And perhaps it would run only a subset on some platforms. I'd also like more robust cleanup and reporting.

gilles-peskine-arm commented:
I've rebased this branch on top of development. The old branch is at https://github.com/gilles-peskine-arm/mbedtls/tree/all.sh-keep-going-1

I added an extra commit with a bit of documentation.

Jenkins release job: https://jenkins-internal.mbed.com/job/mbedtls-release/68/

mpg commented on tests/scripts/all.sh:
```diff
@@ -4,20 +4,78 @@
+#   * GCC and Clang (recent enough for using ASan with gcc and MemSan with clang, or valgrind)
+#   * arm-gcc and mingw-gcc
+#   * ArmCC 5 and ArmCC 6, unless invoked with --no-armcc
+#   * , unless invoked with --no-yotta
```
Missing word: yotta

gilles-peskine-arm replied:
Damn, you're too fast. I just force-pushed a commit with the missing word. Sorry.

gilles-peskine-arm commented:
`all.sh --no-armcc` still runs `armcc` from `output_env.sh`. This still works even if you don't have armcc, or if you do but no license is available. When I run it with no network connection, armcc tries to contact the license server but this takes <1s so I think this is acceptable. I'd rather not add machinery to pass options to `output_env.sh` at this stage.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -56,28 +116,38 @@ fi
-    make clean
+    command make clean
 
     find . -name yotta -prune -o -iname '*cmake*' -not -name CMakeLists.txt -exec rm -rf {} \+
```
Should be skipped if `[ YOTTA -eq 0 ]`.

mpg replied:
The only yotta-related thing here is we ignore a directoty named `yotta` if one is present. I think it's an ok thing to do even if yotta testing is disabled.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -195,7 +299,7 @@ if [ $FORCE -eq 1 ]; then
     cleanup
```
Above, `[ $YOTTA -eq 1 ]` should be `[ $YOTTA -ne 0 ]` for consistency.

gilles-peskine-arm commented:
`tests/scripts/all.sh --no-armcc` passed locally while no license for armcc was available.
